### PR TITLE
Hotfix/cbse 8134

### PR DESCRIPTION
**Pull request by bmeike**
Attempt to make it more difficult to release the document underneath an array/dict iterator.

It turns out that MCollections have a ref to the C4Document into which they point, in their context.  That means the C4Document won't be GCed while the collection is around.  If the iterators keep references to their MCollections, maybe RefCounted is not necessary.

### CODE REVIEW

File: src/main/java/com/couchbase/lite/internal/core/RefCounted.java [hunk 1]
```diff
@@ -27,7 +27,7 @@ public synchronized void retain() {
     }
 
     public synchronized void release() {
-        if (--refCount <= 0) { free(); }
+        // if (--refCount <= 0) { free(); }
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/internal/core/RefCounted.java:30]: The release() is doing nothing now. We don't need this anymore?

Comment (bmeike) [reply src/main/java/com/couchbase/lite/internal/core/RefCounted.java:30]: I think that is true.  Here is my reasoning:

I think that the intention was that call to retain in the `DocContext` would hold the `c4Document` until the Context was no longer needed.  That just doesn't work, though, because there is also a call to `free` in the `AbstractDocument`'s finalizer.  So, the only thing that this `release` call could *ever* do, is to release the document too early.

I am still worried about escape analysis.  It seems quite likely to me that, because there is no way to refer to the c4Document in the DocContext, it will be GCed practically immediately.

Also, FWIW, it is possible to change an `AbstractDocument`'s `c4Document`.  I am not at all convinced that the DocContext points at the right c4Document, if that happens.

Comment (pasin) [reply src/main/java/com/couchbase/lite/internal/core/RefCounted.java:30]: This is very hard to follow. Probably we could just remove RefCounted class.

Comment (bmeike) [reply src/main/java/com/couchbase/lite/internal/core/RefCounted.java:30]: This is a VF on a release branch.  Changes should be minimal.

Comment (bmeike) [reply src/main/java/com/couchbase/lite/internal/core/RefCounted.java:30]: ... and, btw, it *is* gone in 2.8

**Comment by bmeike** [comment]
This branch will not be merged.  It will be used to build the VF

**Comment by bmeike** [comment]
Released as 2.6.4.2
